addons/product_images/wizard/product_fetch_image_wizard.py

```diff
@@ -0,0 +1,243 @@
+            )
+        else:
+            if len(products_to_process) > 100:
+                self.env.ref('product_images.ir_cron_fetch_image')._trigger(
+                    fields.Datetime.now() + timedelta(minutes=1.0)
+                )
+
+    def _fetch_image_from_google(self, products):
+        """ Fetch images of a list of products from Google
+
+        Retrieve an image for each product in `products` from the Google Custom Search API
+        using their barcode. Fetch the 10 first images url and save the first valid image.
+
+        :param products: Product variant recordset
+        """
+        if not products:
+            return
+
+        ICP = self.env['ir.config_parameter']
+        session = requests.Session()
+        custom_search_req = requests.Request(
+            method='GET',
+            url='https://customsearch.googleapis.com/customsearch/v1',
+            params={
+                'cx': ICP.get_param('google.pse.id').strip(),
+                'safe': 'active',
+                'searchType': 'image',
+                'key': ICP.get_param('google.custom_search.key').strip(),
+                'rights': 'cc_publicdomain,cc_attribute,cc_sharealike',
+                'imgSize': 'large',
+                'imgType': 'photo',
+                'fields': 'searchInformation/totalResults,items(link)',
+            },
+        )
+        prepared_req = session.prepare_request(custom_search_req)
+        request_url = prepared_req.url
+        for product in products:
+            try:
+                prepared_req.prepare_url(request_url, {'q': product.barcode})
+
+                response = session.send(prepared_req, timeout=5)
+                response.raise_for_status()
+                response_content = response.json()
+            except (ConnectTimeout, ReadTimeout):
+                continue
+            except HTTPError as http_error:
+                if http_error.response.status_code == requests.codes.service_unavailable:
+                    continue
+                else:
+                    raise
+            else:
+                if int(response_content['searchInformation']['totalResults']) > 0:
+                    for item in response_content['items']:
+                        if 'x-raw-image:///' in item['link']:  # Ignores images with an incorrect link
+                            continue
+                        try:
+                            req = session.get(item['link'], timeout=5)
+                            req.raise_for_status()
+                            if 'image/' in req.headers['Content-Type']:
+                                product.image_1920 = base64.b64encode(req.content)
+                                break
+                        except Exception:
+                            continue
+            finally:
+                product.image_fetch_pending = False
+                self.env.cr.commit()  # Commit every image in case the cron is killed
```

Comment:
I'm really not sure about this. What bothers me is that:
- you silently ignore lots of exceptions that won't reach the caller (cron or action).
- you ignore them for no apparent reason -> if it's not obvious, comment!
- instead of handling exceptions either in the caller or the callee, they are handled in both. For example, `except HTTPError` can be found in 3 different places in this file.
- `except Exception: continue` is the best way to be totally unaware of potential errors.

Reply:
I've added comments to the most ambiguous exception handling in this function.

For the exceptions handled in the caller and the callee (e.g. `HTTPError`), they're not the same type of error. In `_fetch_image_from_google`, the `HTTPError` is a service_unavailble error that is raised when Google has an internal issue. The `HTTPError` in `_cron_fetch_image` and `action_fetch_image` are similar because we handle the same error from the cron and the action button and are necessary because that the cron is executed in the background and don't enable the same interaction with the user as an action button.